planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+
+
+## Project / Assembly Restructuring
+
+* Pull everything to do with Schema object generation and difference detection to a separate library (`IFeatureSchema`, `ISchemaObject`, etc.). Mostly to clean out the main library, but also because this code could easily be reused outside of Marten. Separating it out might make it easier to test and extend that functionality, which is something that occasionally gets requested. There's also the possibility of further breaking that into abstractions and implementations for the long run of getting us ready for Sql Server or other database engine support
+
+* *Possibly* pull the ADO.Net helper code like `CommandBuilder` and the extension methods into a small helper library somewhere else (I'm nominating the [Baseline](https://jasperfx.github.io/baseline) repository). This code is copied around to other projects as it is, and it's another way of getting stuff out of the main library and the test suite.
+
+* More in the event store section, but let's get the event store functionality into its own library (or possibly a couple other libraries for async projection support).
+
+## F# Improvements
+
+WE NEED TO HAVE AN F# SUBCOMMITTEE TO HELP HERE. 
+
+## HostBuilder Integration
+
+If he's okay with this, I'd vote to bring Joona-Pekka Kokko's ASP.Net Core integration library into the main repository and make that the officially blessed and documented recipe for integrating Marten into .Net Core applications based on the `HostBuilder` in .Net Core. I suppose we could also multi-target `IWebHostBuilder` for ASP.Net Core 2.*.
+
+That `HostBuilder` integration could be extended to:
+
+* Optionally set up the Async Daemon in an `IHostedService` -- more on this in the Event Store section
+* Optionally register some kind of `IDocumentSessionBuilder` that could be used to customize session construction? 
+* Have some way to have container resolved `IDocumentSessionListener` objects attached to `IDocumentSession`. This is to have an easy recipe for folks who want events broadcast through messaging infrastructure in CQRS architectures
+
+## Command Line Support
+
+The `Marten.CommandLine` package already uses Oakton for command line parsing. For easier integration in .Net Core applications, we could shift that to using the [Oakton.AspNetCore](https://jasperfx.github.io/oakton/documentation/aspnetcore/) package so the command line support can be added to any ASP.net Core 2.* or .Net Core 3.* project by installing the Nuget. This might simplify the usage because you'd no longer need a separate project for the command line support.
+
+There are some long standing stories about extending the command line support for the event store projection rebuilding. I think that would be more effective if it switches over to Oakton.AspNetCore.
+
+
+## Linq
+
+This is also covered by the [Linq Overhaul](https://github.com/JasperFx/marten/issues/1201) issue.
```

Comment:
I agree with the points below. For me, the most important is to have good support for the nested object and collections as that's one of the biggest benefits of using the document database vs ORM.

Reply:
We'll have to spike this out, but I *think* we could maybe be using common table expressions to help out with queries within collections. Maybe.

Reply:
*Maybe* the common table expression stuff will help here. I'm curious to try to start using JsonPath with a PLV8 / JS polyfill for < 12 if we have to.

Reply:
The idea with the polyfill sounds intriguing 👍